File: array.cpp
#include "array.h"
#include <iostream>

void CreateMArray(MArray* arr, int capacity) {
    arr -> size = 0;
    arr -> capacity = capacity;
    arr -> data = new string[arr -> capacity];
}

void ResizeMArray(MArray* arr) {
    int newCapacity = arr -> capacity * 2;
    string* newData = new string[newCapacity];
    for (int i = 0; i < arr -> size; i++) {
        newData[i] = arr -> data[i];
    }
    delete[] arr -> data;
    arr -> data = newData;
    arr -> capacity = newCapacity;
}

void MPushBack(MArray* arr, string element) {
    if (arr -> size == arr -> capacity) ResizeMArray(arr);
    arr -> data[arr -> size] = element;
    arr -> size++;
}

void MPushIndex(MArray* arr, string element, int index) {
    if (index < 0 || index > arr -> size) return;
    if (arr -> size >= arr -> capacity) ResizeMArray(arr);
    for (int i = arr -> size; i > index; i--) {
        arr -> data[i] = arr -> data[i - 1];
    }
    arr -> data[index] = element;
    arr -> size++;
}

string MGet(MArray* arr, int index) {
    if (index < 0 || index >= arr -> size) return "";
    return arr -> data[index];
}

void MPopIndex(MArray* arr, int index) {
    if (arr -> size == 0 || index < 0 || index >= arr -> size) return;
    for (int i = index; i < arr -> size - 1; i++) {
        arr -> data[i] = arr -> data[i + 1];
    }
    arr -> size--;
}

void MReplace(MArray* arr, string element, int index) {
    if (index < 0 || index >= arr -> size) return;
    arr -> data[index] = element;
}

int MGetSize(MArray* arr) {
    return arr -> size;
}

void PrintMArray(MArray* arr) {
    cout << "[";
    for (int i = 0; i < arr -> size; i++) {
        cout << arr -> data[i];
        if (i < arr -> size - 1) cout << ", ";
    }
    cout << "]" << endl;
}

void DestroyMArray(MArray* arr) {
    if (arr) delete[] arr -> data;
}

File: array.go
package main

import "fmt"

type MArray struct {
	Data     []string
	Size     int
	Capacity int
}

func CreateMArray(capacity int) *MArray {
	if capacity <= 0 {
		capacity = 10
	}
	return &MArray{
		Data:     make([]string, capacity),
		Size:     0,
		Capacity: capacity,
	}
}

func (arr *MArray) Resize() {
	newCapacity := arr.Capacity * 2
	newData := make([]string, newCapacity)
	copy(newData, arr.Data[:arr.Size])
	arr.Data = newData
	arr.Capacity = newCapacity
}

func (arr *MArray) PushBack(element string) {
	if arr.Size == arr.Capacity {
		arr.Resize()
	}
	arr.Data[arr.Size] = element
	arr.Size++
}

func (arr *MArray) PushIndex(element string, index int) {
	if index < 0 || index > arr.Size {
		return
	}
	if arr.Size >= arr.Capacity {
		arr.Resize()
	}
	for i := arr.Size; i > index; i-- {
		arr.Data[i] = arr.Data[i-1]
	}
	arr.Data[index] = element
	arr.Size++
}

func (arr *MArray) Get(index int) string {
	if index < 0 || index >= arr.Size {
		return ""
	}
	return arr.Data[index]
}

func (arr *MArray) PopIndex(index int) {
	if arr.Size == 0 || index < 0 || index >= arr.Size {
		return
	}
	for i := index; i < arr.Size-1; i++ {
		arr.Data[i] = arr.Data[i+1]
	}
	arr.Size--
}

func (arr *MArray) Replace(element string, index int) {
	if index < 0 || index >= arr.Size {
		return
	}
	arr.Data[index] = element
}

func (arr *MArray) GetSize() int {
	return arr.Size
}

func (arr *MArray) Print() {
	fmt.Print("[")
	for i := 0; i < arr.Size; i++ {
		fmt.Print(arr.Data[i])
		if i < arr.Size-1 {
			fmt.Print(", ")
		}
	}
	fmt.Println("]")
}